[no-merge] Check unused function params

som-snytt opened the pull request:
Both for explicit anon funs and those introduced
by for expr desugarings.

[PR status: improve warning message]

som-snytt committed:
Check unused function params

Both for explicit anon funs and those introduced
by for expr desugarings.

hrhino commented on test/files/neg/warn-unused-params.check:
```diff
@@ -13,6 +13,12 @@ case class CaseyAtTheBat(k: Int)(s: String)        // warn
+            ^
+warn-unused-params.scala:78: warning: parameter value i in value $anonfun is never used
+  val g = for (i <- List(1)) yield 42
+               ^
```
```scala
scala> val f = (_: Int) => 1
<console>:11: warning: parameter value x$1 in value $anonfun is never used
       val f = (_: Int) => 1
                 ^
```

I suggest that we shouldn't warn on unnamed (`WILDCARD`-named) params.

som-snytt replied:
Good catch.

hrhino commented on test/files/neg/warn-unused-params.check:
```diff
@@ -13,6 +13,12 @@ case class CaseyAtTheBat(k: Int)(s: String)        // warn
 warn-unused-params.scala:62: warning: parameter value readResolve in method f is never used
   def f(readResolve: Int) = 42           // warn
         ^
+warn-unused-params.scala:76: warning: parameter value i in value $anonfun is never used
```
... "in this anonymous function is never used" if `p.owner.isAnonymousFunction`?

som-snytt replied:
I'll wait for final tweaking. The goal for the next push is to track usage during `for` desugaring or caramelization.

som-snytt replied:
Done on linked PR.

hrhino reviewed (commented):
More pedantry is always... something.

Semi-relatedly, I turned on only `-Ywarn-unused:explicits` on my company's codebase yesterday; only a thousand lines needed changed. It's for the best.

som-snytt committed:
No warn underscore in anon funs

Placeholder syntax results in synthetic params which
are excluded from unused warnings.

Underscore function params also get a fresh name,
but since they are not synthetic, they receive
the `NoWarnAttachment` which is awkwardly
propagated.

som-snytt commented:
Forward ported, so this should be marked no-merge.

hrhino commented:
does this mean that #6924 is also good to go? It would be a real pity if this got lost somehow.

xuwei-k commented:
regression?  🤔  https://github.com/scala/bug/issues/11175
